jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java

```diff
@@ -1267,25 +1277,24 @@ else if (_endOfContent == EndOfContent.UNKNOWN_CONTENT)
                                     _header = cachedField.getHeader();
                                     _headerString = n;
 
-                                    if (v == null)
+                                    int pos = buffer.position() + n.length() + (v == null ? 0 : v.length()) + 1;
+                                    if (v == null || pos >= buffer.limit())
                                     {
                                         // Header only
                                         setState(FieldState.VALUE);
                                         _string.setLength(0);
                                         _length = 0;
-                                        buffer.position(buffer.position() + n.length() + 1);
+                                        buffer.position(v == null ? pos : pos - v.length());
                                         break;
                                     }
```

Comment:
Does not feel right.
`pos` may be set past the cached value, and we may still enter the `if` block due to `pos >= buffer.limit()`, but then we set the state to be `VALUE`?

Even if it is correct it's just way more difficult to read, so perhaps needs comments and/or examples.

Reply:
restructured to `posAfterName` and `posAfterValue`... but doing some more checks.

Reply:
more cleanup and testing done. OK now.